pkg/cli/cliexecuter.go

```diff
@@ -1,97 +1,263 @@
 
-	return &executer{
-		log:       logger,
-		config:    config,
-		clientSet: clientSet,
+	return &Executer{
+		Client:   client,
+		log:      logger,
+		routines: make(map[types.UID]*routine),
+		runner:   fn,
 	}, nil
 }
 
-// Execute builds and runs a Post request on contianer for SPDY (shell) executor
-func (e executer) Execute(pod *corev1.Pod, command []string) (stdout string, stderr string, err error) {
-	if len(pod.Spec.Containers) == 0 {
-		err := errors.New("create cli executer failed")
-		e.log.Error(err, "No container found in Pod", "Pod Name", pod.Name)
-		return "", "", err
+// AsyncExecute runs the command in a goroutine mapped to the UID
+func (e *Executer) AsyncExecute(ctx context.Context, uid types.UID, command []string, retryCount int, retryInterval, timeout time.Duration) {
+	e.routinesLock.Lock()
+	defer e.routinesLock.Unlock()
+	if _, exist := e.routines[uid]; exist {
+		return
 	}
 
-	var (
-		stdoutBuf bytes.Buffer
-		stderrBuf bytes.Buffer
-	)
-
-	//containerName := pod.Spec.Containers[0].Name
-	containerName := "manager"
-
-	// Build the Post request for SPDY (shell) executor
-	req := e.clientSet.CoreV1().RESTClient().Post().
-		Resource("pods").
-		Name(pod.Name).
-		Namespace(pod.Namespace).
-		SubResource("exec").
-		Param("container", containerName)
-
-	req.VersionedParams(&corev1.PodExecOptions{
-		Container: containerName,
-		Command:   command,
-		Stdin:     false,
-		Stdout:    true,
-		Stderr:    true,
-		TTY:       false,
-	}, scheme.ParameterCodec)
-
-	execSPDY, err := remotecommand.NewSPDYExecutor(e.config, "POST", req.URL())
-	if err != nil {
-		e.log.Error(err, "failed building SPDY (shell) executor")
-		return "", "", err
+	// create a context for the fence agent command that the controller can cancel
+	cancellableCtx, cancel := context.WithCancel(ctx)
+	routine := routine{
+		cancel: cancel,
 	}
+	e.routines[uid] = &routine
 
-	// Execute the Post request for SPDY (shell) executor
-	err = execSPDY.Stream(remotecommand.StreamOptions{
-		Stdout: &stdoutBuf,
-		Stderr: &stderrBuf,
-		Tty:    false,
-	})
-	if err != nil {
-		e.log.Error(err, "Failed to run exec command", "stdout", stdoutBuf.String(), "stderr", stderrBuf.String())
+	go e.fenceAgentRoutine(cancellableCtx, uid, command, retryCount, retryInterval, timeout)
+}
+
+func (e *Executer) fenceAgentRoutine(ctx context.Context, uid types.UID, command []string, retryCount int, retryInterval, timeout time.Duration) {
+	// run the command and update the status
+	retryErr, cmdErr := e.runWithRetry(ctx, uid, command, retryCount, retryInterval, timeout)
+	if retryErr != nil {
+		switch {
+		case errors.Is(retryErr, context.Canceled):
+			e.log.Info("fence agent context canceled. Nothing to do")
+			return
+		case wait.Interrupted(retryErr):
+			e.log.Info("fence agent context timed out")
+		default:
+			e.log.Error(retryErr, "fence agent retry error")
+		}
+	}
+
+	retryErr = e.updateStatusWithRetry(ctx, uid, cmdErr)
```

Comment:
It can also be the retry error, but since it's just an error and it is not used elsewhere, I can replace it with 

```
if err := e.updateStatusWithRetry(...); err != nil {}
```